airflow/models/taskinstance.py

```diff
@@ -2393,6 +2393,12 @@ def _run_raw_task(
                 self.handle_failure(e, test_mode, context, session=session)
                 session.commit()
                 raise
+            except SystemExit as code:
+                self.handle_failure(
+                    f"Task failed due to SystemExit({code})", test_mode, context, session=session
+                )
+                session.commit()
+                raise
```

Comment:
I wonder if we should exclude the exit code 0 and set the state to success 🤔

Reply:
yep . And actually we should handle None here as well in the same way https://docs.python.org/3/library/exceptions.html#SystemExit

Reply:
Usually, we don't use the exit(0) or exit(), to gracefully finish the application. If we use it means, we want to abruptly exit the application. I believe this only applies to the Python operators, not to the other operators.

Reply:
I suppose all other operators, in which we don't allow client code to execute as part of Airflow processes, must use Airflow's exceptions like AirflowException. This way, the only question is if it is okay to use `sys.exit()` or `sys.exit(0)` in the client code or not. I do not have an answer, but one thing worth considering is if we consider zero (or None) code as success, we could break backward compatibility because now such tasks are marked as failed with zombie detection

Reply:
It's not the matter if we use it or not. But our users can choose to do so. Airflow task is very special in the sense that it has to be rather resilient to anything our (often in a very inventive way) do in their tasks. If someone chooses to do `sys.system.exit(0)` in their PythonOperator's callable to finish their task quickly, they are perfectly OK to do so and we cannot do anything about it. 

And if we handle it properly in this try/finally this is also perfectly valid from Airflow point of view. There is a reason why `sys.exit()` - works the way it works, by raising SystemExit exception. This is done specifically to have all the try/finally clauses to be able to finalize whatever they were supposed to finalize if the code was called from "somewhere else" (this is o top of the fact that we are also forking when executing tasks but this is a differnet story for another day). 

Raising AirflowException is neither forced not even required, nor AirflowException is handled in any other way than any other exception. It's merely a base class that all our "specialized" exceptions derive from (for example AirflowSkipException when raised causing the task to be skipped or AirflowFailException that causes task to not retry. But there is nowhere expectation or requirement that user's code must or even should try AirflowException in case oi error. It can throw any kind of exception or it can call sys.exit(1) . Similarly there is no strong requirement that `execute()` method  returns "as usual". Yes, if you want to return some value (and then store it as xcom automatically), surely the best way is to just return it from the `execute()` method (BTW. you can also manually save the xcom value manually and then sys.exit(0) so this is also not the only way). But if you do not want to do it, doiing sys.exit(0) is perfectly good way for the users to finish their task and signal success. 

BTW. Side commen. There is this great `core.py` podcast by Łukasz Langa and Pablo Galindo and the last episode (whicih I just started listening) called "The Exceptional Episode" is all about that - how exceptions are handled in Python. On spotify here https://open.spotify.com/episode/76eZ8LVhzQLHrS2GRhOjG7.

Reply:
>  because now such tasks are marked as failed with zombie detection

I think they are not. Where do you have the notion of that from?

Reply:
> > because now such tasks are marked as failed with zombie detection
> 
> I think they are not. Where do you have the notion of that from?

You can check it in the main branch with this simple DAG:

```python
from datetime import datetime, timedelta

from airflow import DAG
from airflow.decorators import task

with DAG(
    "zombie_detection_check",
    schedule_interval="@daily",
    start_date=datetime.now() - timedelta(days=2),
):
    @task
    def zombie():
        exit(-1)
    zombie()

    @task
    def zombie_without_exit_code():
        exit()
    zombie_without_exit_code()
```

To disable zombie detection, you can add a return statement to `_find_zombies` in `scheduler_job_runner.py` before running the DAG:

```python
    # [START find_zombies]
    def _find_zombies(self) -> None:
        return
```

After doing this, you will observe tasks stacked in the RUNNING state.

![image](https://github.com/apache/airflow/assets/54231417/5ea986c8-5f69-4f67-b3aa-93d7297fbecf)

Reply:
Then we shoudl fix it.

Reply:
And just to add- likey somewhere in our stack we do not save the state of task in "finally" clause - that's the most probable reason why it happens. I'd say it falls in this PR rather nicely.

Reply:
> Then we should fix it.

Of course! This PR aims to do just that.

I've made adjustments. And now handling of system exit in tasks works like this:
* Exiting with a zero or None code signifies success, and the task does not return any value.
* Exiting with other codes signifies an error.

@potiuk @dirrao @hussein-awala, can you take a look please?